src/Razor/src/Microsoft.AspNetCore.Razor.Language/RazorCodeDocumentExtensions.cs

```diff
@@ -208,48 +212,129 @@ internal static bool TryComputeNamespaceAndClass(this RazorCodeDocument document
+            {
+                lastNamespaceContent = namespaceContent;
+                lastNamespaceLocation = namespaceLocation;
+            }
+
+            // If there are multiple @namespace directives in the heirarchy,
+            // we want to pick the closest one to the current document.
+            if (!string.IsNullOrEmpty(lastNamespaceContent))
+            {
+                baseNamespace = lastNamespaceContent;
+                var directiveLocationDirectory = NormalizeDirectory(lastNamespaceLocation.FilePath);
+
+                // We're specifically using OrdinalIgnoreCase here because Razor treats all paths as case-insensitive.
+                if (!document.Source.FilePath.StartsWith(directiveLocationDirectory, StringComparison.OrdinalIgnoreCase) ||
+                    document.Source.FilePath.Length <= directiveLocationDirectory.Length)
+                {
+                    // The imports are not from the directory hierarchy, can't compute a suffix.
+                    appendSuffix = false;
+                }
+                else
+                {
+                    // We know that the document containing the namespace directive is in the current document's heirarchy.
+                    // Let's compute the actual relative path that we'll use to compute the namespace suffix.
+                    relativePath = document.Source.FilePath.Substring(directiveLocationDirectory.Length);
+                }
+            }
+            else
+            {
+                var options = document.GetCodeGenerationOptions() ?? document.GetDocumentIntermediateNode()?.Options;
+                var rootNamespace = options?.RootNamespace;
+                if (string.IsNullOrEmpty(rootNamespace))
+                {
+                    // Couldn't compute namespace in any way.
+                    @namespace = null;
+                    return false;
+                }
+
+                baseNamespace = rootNamespace;
+                appendSuffix = true;
             }
 
             var builder = new StringBuilder();
 
             // Sanitize the base namespace, but leave the dots.
-            var segments = rootNamespace.Split(NamespaceSeparators, StringSplitOptions.RemoveEmptyEntries);
+            var segments = baseNamespace.Split(NamespaceSeparators, StringSplitOptions.RemoveEmptyEntries);
             builder.Append(CSharpIdentifier.SanitizeIdentifier(segments[0]));
             for (var i = 1; i < segments.Length; i++)
             {
                 builder.Append('.');
                 builder.Append(CSharpIdentifier.SanitizeIdentifier(segments[i]));
             }
 
-            segments = relativePath.Split(PathSeparators, StringSplitOptions.RemoveEmptyEntries);
-
-            // Skip the last segment because it's the FileName.
-            for (var i = 0; i < segments.Length - 1; i++)
+            if (appendSuffix)
```

Comment:
I'd probably rename this to `appendPathToNamespace` or something else more descriptive